### PR TITLE
Add the numpy implementation in the keras backend documentation

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
### Summary

Since we have implemented functions in numpy which are equivalent of the keras abstract backend, it is a very good opportunity to improve the documentation by using that. Here is what I mean:

gather
reverse

The code displayed in the documentation is being automatically taken from `numpy_backend.py`

### PR Overview

I had to rename and move `reference_operations.py` to avoid having to import a python file by path (very ugly).

Other than that the code is pretty simple and self explanatory.

- [ ] This PR requires new unit tests [y/n] (make sure tests are included)
- [ ] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by fchollet** [comment]
This is a great idea.

Are there any ops for which the Numpy version would be too long or hard to read, and thus should be hidden behind a toggle "show Numpy implementation"?

**Comment by gabrieldemarmiesse** [comment]
Yes, there are. This is why for this first PR, I've only included simple functions. The pooling operations and convolutions have 30+ LOC. 

A `show numpy implementation` would be super nice but I didn't find how to do that in mkdocs. I don't know if hidden text is supported.

**Comment by fchollet** [comment]
You can insert arbitrary HTML into md files. We would just need to create a JS script for this purpose (could be just a couple of lines).

File: docs/autogen.py [hunk 1]
```diff
@@ -81,6 +81,7 @@
 from keras import losses
 from keras import metrics
 from keras import backend
+from keras.backend import numpy_backend as KNP
```
Comment (fchollet) [inline docs/autogen.py]: Prefer naming it numpy_backend for clarity.

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -1284,6 +1284,11 @@ def gather(reference, indices):
 
     # Returns
         A tensor of same type as `reference`.
+
+    # Numpy implementation
+    ```python
+        {np_implementation}
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Use `{{}}` for templating markers

**Comment by taehoonlee** [comment]
Good idea, @gabrieldemarmiesse! How about inserting the header and the footer into `np_impl_code`? My suggestion is:
 - as-is:
```
    # Numpy implementation
    ```python
        {{np_implementation}}
    ```
```
 - to-be:
```
    {{np_implementation}}
```

**Comment by gabrieldemarmiesse** [comment]
@taehoonlee thanks.

Indeed, what you are saying makes sense. I'm in favor of that. The only thing is that we won't be able to add custom text before and after the code. But I suppose it's fine since we can put comments in the code. Let me change that.

**Comment by gabrieldemarmiesse** [comment]
I suggest that we merge this, because this is an easy win for the documentation. 

For the long implementations, let's not include them now and wait until a PR is done with the corresponding javascript. (help welcome since I haven't done any javascript and html for a while now).

**Comment by gabrieldemarmiesse** [comment]
Actually that was easy to add a show button.

In this implementation, if the function has more than 10 LOC, a button `Show the numpy implementation` is created automatically.

**Comment by gabrieldemarmiesse** [comment]
Note: this was only tested with firefox and chrome. I don't know if other browsers support those tags.

**Comment by fchollet** [comment]
Please fix PEP8 issues: https://travis-ci.org/keras-team/keras/jobs/448960181

**Comment by fchollet** [comment]
One thing to keep in mind is that this makes the Numpy backend a part of the documentation, and thus user-facing (not as an API, but in terms of code-as-documentation). This sets a high standard of readability and clarity for any code that goes into the Numpy backend.

**Comment by gabrieldemarmiesse** [comment]
Ok I'll keep that in mind for any function that we choose to display in the docs.

**Pull request merged**